Revert add parent img refcount for faster rmi

coolljt0725 opened the pull request:
Signed-off-by: Lei Jitang leijitang@huawei.com
#16818  add parent img refcount for faster rmi

but it can't work for the exist images and make some bad regression.
This will make the existed images can be deleted even it has children
and make `docker image` show all the images without `-a`
The reason is that `parentRefs` is only initialize by `Register`, and 
`Register` is only be called when create a new image by such as `docker pull/load/commit`
but the already exists images don't.

For now, it's better to revert this until we found a better way.

coolljt0725 committed:
Revert add parent img refcount for faster rmi

Signed-off-by: Lei Jitang <leijitang@huawei.com>

runcom commented:
Cool, thanks for this @coolljt0725, I'll work to see what I can do to fix this, sadly there were no tests for this :( I'll add them once done

coolljt0725 commented:
@runcom  Maybe we should call `Register` during daemon starting just like we do on registering containers.  :)

runcom commented:
@coolljt0725 maybe, btw I can't reproduce the rmi of a Parent image

runcom commented:
```
$ docker inspect ubuntu | grep -i parent
    "Parent": "d74508fb6632491cea586a1fd7d748dfc5274cd6fdfedee309ecdcbc2bf5cb82",
$ docker rmi d74508fb6632491cea586a1fd7d748dfc5274cd6fdfedee309ecdcbc2bf5cb82
Error response from daemon: conflict: unable to delete d74508fb6632 (cannot be forced) - image has dependent child images
Error: failed to remove images: [d74508fb6632491cea586a1fd7d748dfc5274cd6fdfedee309ecdcbc2bf5cb82]
```

runcom commented:
> @runcom Maybe we should call Register during daemon starting just like we do on registering containers. :)

Maybe `NewGraph` needs to `Register` all images in `root` if root exists

coolljt0725 commented:
@runcom  I think if you restart the docker deamon, you can reproduce the rmi of a Parent image
<pre><code>$ docker inspect busybox | grep -i parent
    "Parent": "79722f6accc3345ac95c95dd5dd4af5563cd68a04691e728df4724fc8e8ad581",
$ docker rmi 79722f6accc3345ac95c95dd5dd4af5563cd68a04691e728df4724fc8e8ad581
Deleted: 79722f6accc3345ac95c95dd5dd4af5563cd68a04691e728df4724fc8e8ad581
Deleted: cf2616975b4a3cba083ca99bc3f0bf25f5f528c3c52be1596b30f60b0b1c37ff
</code></pre>

>Maybe NewGraph needs to Register all images in root if root exists

agree :)

runcom commented:
still can't repro (but ok I see it can happen):
```
$ docker inspect busybox | grep -i parent                                  
    "Parent": "cfa753dfea5e68a24366dfba16e6edf573daa447abf65bc11619c1a98a3aff54",
$ docker rmi cfa753dfea5e68a24366dfba16e6edf573daa447abf65bc11619c1a98a3aff54
Error response from daemon: conflict: unable to delete cfa753dfea5e (cannot be forced) - image has dependent child images
Error: failed to remove images: [cfa753dfea5e68a24366dfba16e6edf573daa447abf65bc11619c1a98a3aff54]
```

runcom commented:
btw graph has a `restore` method which should do the trick

runcom commented:
`docker images -a` is not showing every imges btw, it's showing images <none> but they have a parent

coolljt0725 commented:
@runcom  I have made some investigating, with this patch can solve the problem.
<pre><code>$ git diff
diff --git a/graph/graph.go b/graph/graph.go
index 7932fa2..c9f20fe 100644
--- a/graph/graph.go
+++ b/graph/graph.go
@@ -148,6 +148,22 @@ func (graph *Graph) restore() error {
                id := v.Name()
                if graph.driver.Exists(id) {
                        ids = append(ids, id)
+                       pth := filepath.Join(graph.root, id, "json")
+                       jsonSource, err := os.Open(pth)
+                       if err != nil {
+                               return err
+                       }
+                       defer jsonSource.Close()
+                       decoder := json.NewDecoder(jsonSource)
+                       img := &image.Image{}
+                       if err := decoder.Decode(img); err != nil {
+                               return err
+                       }
+                       graph.parentRefsMutex.Lock()
+                       if img.Parent != "" {
+                               graph.parentRefs[img.Parent]++
+                       }
+                       graph.parentRefsMutex.Unlock()
                }
        }
</code></pre>

runcom commented:
I'll try that thanks!

runcom commented:
There seems to be some weird stuff with images build from scratch as well :) I'm fixing it btw

coolljt0725 commented:
@runcom  Feel free to close this when you have it fixed. 
it's midnight and I have to rest  :)

runcom commented:
I think this is good to merge till I find all it's going wrong and resubmit a PR with regression tests
ping @LK4D4 @icecrime

LK4D4 commented:
LGTM
